### PR TITLE
HyperFrame

### PR DESCRIPTION
**Pull request by sam-lawrence**

### CODE REVIEW

File: tabulate.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            inner_content_by_path = extract_content_from_tag(inner_element)
+            for inner_path, inner_content in inner_content_by_path.items():
+                element_path = (tag.name, i, *inner_path)
+                content_by_hyperpath[element_path] = inner_content
+
+    return content_by_hyperpath
+
+
+def tabulate_from_hyperpattern_tags(tags_with_structure: list[Tag]) -> pd.DataFrame:
+    tag_contents = [extract_content_from_tag(tag) for tag in tags_with_structure]
```
Comment (risktoparkrageram) [inline tabulate.py]: would be nice to add a docstring about the format of the returned dataframe here

File: hyperpattern.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        """
+        Check if this structure is maximal.
+
+        A structure is not maximal if all of it's tags have different parents
+        but the same parent structure.
+
+        In other words a structure is maximal if there are no larger structures
+        for which there is a one-to-one relationship between instances of the smaller
+        structure and the larger structure defined by the larger instance
+        containing the smaller instance.
```
Comment (risktoparkrageram) [inline hyperpattern.py]: :+1:

**Review by risktoparkrageram — APPROVED** [review]
looks great

File: tabulate.py (hunk 1 above)
Comment (sam-lawrence) [reply tabulate.py]: 15cc75acf9f2d6bcdfbd3a57d40c5ab93e04daf6

**Comment by sam-lawrence** [comment]
Thanks @risktoparkrageram !

**Pull request merged**